File: core/src/main/com/thymeflow/rdf/sail/InterceptingSailConnection.java
package com.thymeflow.rdf.sail;

import org.eclipse.rdf4j.model.IRI;
import org.eclipse.rdf4j.model.Resource;
import org.eclipse.rdf4j.model.Value;
import org.eclipse.rdf4j.sail.NotifyingSailConnection;
import org.eclipse.rdf4j.sail.SailException;
import org.eclipse.rdf4j.sail.UpdateContext;
import org.eclipse.rdf4j.sail.helpers.NotifyingSailConnectionWrapper;

/**
 * @author Thomas Pellissier Tanon
 */
class InterceptingSailConnection extends NotifyingSailConnectionWrapper {

    private NotifyingSailConnection wrappedSailConnection;
    private SailInterceptor interceptor;

    InterceptingSailConnection(NotifyingSailConnection wrappedSailConnection, SailInterceptor interceptor) {
        super(wrappedSailConnection);

        this.wrappedSailConnection = wrappedSailConnection;
        this.interceptor = interceptor;
    }

    @Override
    public void addStatement(UpdateContext updateContext, Resource subj, IRI pred, Value obj, Resource... contexts) throws SailException {
        if (interceptor.onSparqlAddStatement(subj, pred, obj, contexts)) {
            wrappedSailConnection.addStatement(updateContext, subj, pred, obj, contexts);
        }
    }

    @Override
    public void removeStatement(UpdateContext updateContext, Resource subj, IRI pred, Value obj, Resource... contexts)
            throws SailException {
        if (interceptor.onSparqlRemoveStatement(subj, pred, obj, contexts)) {
            wrappedSailConnection.removeStatement(updateContext, subj, pred, obj, contexts);
        }
    }
}

File: core/src/main/com/thymeflow/rdf/sail/InterceptingSailConnection.scala
package com.thymeflow.rdf.sail

import com.thymeflow.rdf.model.{StatementSet, StatementSetDiff}
import org.eclipse.rdf4j.model.{IRI, Resource, Value, ValueFactory}
import org.eclipse.rdf4j.sail.helpers.NotifyingSailConnectionWrapper
import org.eclipse.rdf4j.sail.{NotifyingSailConnection, UpdateContext}

/**
  * @author Thomas Pellissier Tanon
  * @author David Montoya
  */
class InterceptingSailConnection(wrappedSailConnection: NotifyingSailConnection,
                                 valueFactory: ValueFactory,
                                 interceptor: SailInterceptor) extends NotifyingSailConnectionWrapper(wrappedSailConnection) {
  private var addedStatements = StatementSet.empty(valueFactory)
  private var removedStatements = StatementSet.empty(valueFactory)

  override def addStatement(updateContext: UpdateContext, subj: Resource, pred: IRI, obj: Value, contexts: Resource*) {
    addedStatements.add(subj, pred, obj, contexts: _*)
  }

  override def removeStatement(updateContext: UpdateContext, subj: Resource, pred: IRI, obj: Value, contexts: Resource*) {
    removedStatements.add(subj, pred, obj, contexts: _*)
  }

  override def startUpdate(modify: UpdateContext) {
    wrappedSailConnection.startUpdate(modify)
  }

  override def endUpdate(modify: UpdateContext) {
    wrappedSailConnection.endUpdate(modify)
    val diff = StatementSetDiff(addedStatements, removedStatements)
    addedStatements = StatementSet.empty(valueFactory)
    removedStatements = StatementSet.empty(valueFactory)
    interceptor.onSparqlUpdateDiff(diff)
  }
}